ul/src/pdu/reader.rs

```diff
@@ -1,115 +1,13 @@
-    InvalidMaxPdu {
-        max_pdu_length: u32,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("No PDU available"))]
-    NoPduAvailable { backtrace: Backtrace },
-
-    #[snafu(display("Could not read PDU"))]
-    ReadPdu {
-        source: std::io::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Could not read PDU item"))]
-    ReadPduItem {
-        source: std::io::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Could not read PDU field `{}`", field))]
-    ReadPduField {
-        field: &'static str,
-        source: std::io::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display("Invalid item length {} (must be >=2)", length))]
-    InvalidItemLength { length: u32 },
-
-    #[snafu(display("Could not read {} reserved bytes", bytes))]
-    ReadReserved {
-        bytes: u32,
-        source: std::io::Error,
-        backtrace: Backtrace,
-    },
-
-    #[snafu(display(
-        "Incoming pdu was too large: length {}, maximum is {}",
-        pdu_length,
-        max_pdu_length
-    ))]
-    PduTooLarge {
-        pdu_length: u32,
-        max_pdu_length: u32,
-        backtrace: Backtrace,
-    },
-    #[snafu(display("PDU contained an invalid value {:?}", var_item))]
-    InvalidPduVariable {
-        var_item: PduVariableItem,
-        backtrace: Backtrace,
-    },
-    #[snafu(display("Multiple transfer syntaxes were accepted"))]
-    MultipleTransferSyntaxesAccepted { backtrace: Backtrace },
-    #[snafu(display("Invalid reject source or reason"))]
-    InvalidRejectSourceOrReason { backtrace: Backtrace },
-    #[snafu(display("Invalid abort service provider"))]
-    InvalidAbortSourceOrReason { backtrace: Backtrace },
-    #[snafu(display("Invalid presentation context result reason"))]
-    InvalidPresentationContextResultReason { backtrace: Backtrace },
-    #[snafu(display("invalid transfer syntax sub-item"))]
-    InvalidTransferSyntaxSubItem { backtrace: Backtrace },
-    #[snafu(display("unknown presentation context sub-item"))]
-    UnknownPresentationContextSubItem { backtrace: Backtrace },
-    #[snafu(display("Could not decode text field `{}`", field))]
-    DecodeText {
-        field: &'static str,
-        #[snafu(backtrace)]
-        source: dicom_encoding::text::DecodeTextError,
-    },
-    #[snafu(display("Missing application context name"))]
-    MissingApplicationContextName { backtrace: Backtrace },
-    #[snafu(display("Missing abstract syntax"))]
-    MissingAbstractSyntax { backtrace: Backtrace },
-    #[snafu(display("Missing transfer syntax"))]
-    MissingTransferSyntax { backtrace: Backtrace },
-}
-
-pub type Result<T> = std::result::Result<T, Error>;
+pub type Result<T> = std::result::Result<T, ReadError>;
 
-pub fn read_pdu<R>(reader: &mut R, max_pdu_length: u32, strict: bool) -> Result<Pdu>
-where
-    R: Read,
+pub fn read_pdu(mut buf: impl Buf, max_pdu_length: u32, strict: bool) -> Result<Option<Pdu>>
```

Comment:
The changes here are basically just using the methods from `Buf` instead of the methods from the `ReadExt` trait from the `byteorder` crate.  Ended up removing `read_n` since the `copy_to_bytes` method basically does that